### DISCUSSION TITLE
[help] Pipeline error traces to `tar_make` instead of a job, makes debugging hard

### DISCUSSION THREAD

**Discussion by dshemetov** [ropensci/targets#1250] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hi there, we're having a mysterious error in our pipeline and just wanted to see if any of the errors here could give us a hint for where to look (it's a large pipeline so it takes a long time to debug). What's confusing here is that the error message isn't traced to a particular job, but to `tar_make` itself. The line "received data could not be converted to integer" might be informative, but that string doesn't appear in `targets` or `crew`. 

```r
# Lots of dispatched/startup messages prior
# ...
▶ dispatched target forecast_by_ahead_nonmystic.teary.14
received data could not be converted to integer
✔ skipped pipeline [11.184 minutes]
There were 50 or more warnings (use warnings() to see the first 50)
Error:
! Error running targets::tar_make()
Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
Debugging guide: https://books.ropensci.org/targets/debugging.html
How to ask for help: https://books.ropensci.org/targets/help.html
Last error message:
    dims [product 315] do not match the length of object [1]
Last error traceback:
    base::tryCatch(base::withCallingHandlers({ NULL base::saveRDS(base::do.c...
    tryCatchList(expr, classes, parentenv, handlers)
    tryCatchOne(tryCatchList(expr, names[-nh], parentenv, handlers[-nh]), na...
    doTryCatch(return(expr), name, parentenv, handler)
    tryCatchList(expr, names[-nh], parentenv, handlers[-nh])
    tryCatchOne(expr, names, parentenv, handlers[[1L]])
    doTryCatch(return(expr), name, parentenv, handler)
    base::withCallingHandlers({ NULL base::saveRDS(base::do.call(base::do.ca...
    base::saveRDS(base::do.call(base::do.call, base::c(base::readRDS("/tmp/R...
    base::do.call(base::do.call, base::c(base::readRDS("/tmp/RtmpGb0jWW/call...
    (function (what, args, quote = FALSE, envir = parent.frame()) { if (!is....
    (function (targets_function, targets_arguments, options, envir = NULL, s...
    tryCatch(out <- withCallingHandlers(targets::tar_callr_inner_try(targets...
    tryCatchList(expr, classes, parentenv, handlers)
    tryCatchOne(expr, names, parentenv, handlers[[1L]])
    doTryCatch(return(expr), name, parentenv, handler)
    withCallingHandlers(targets::tar_callr_inner_try(targets_function = targ...
    targets::tar_callr_inner_try(targets_function = targets_function, target...
    do.call(targets_function, targets_arguments)
    (function (pipeline, path_store, names_quosure, shortcut, reporter, seco...
    crew_init(pipeline = pipeline, meta = meta_init(path_store = path_store)...
    self$run_crew()
    self$iterate()
    self$process_target(queue$dequeue())
    self$run_target(name)
    if_any(target_should_run_worker(target), self$run_worker(target), self$r...
    self$run_worker(target)
    self$controller$push(command = command, data = data, globals = globals, ...
    .subset2(control, "push")(command = command, data = data, globals = glob...
    .subset2(self, "scale")(throttle = throttle)
    private$.launcher$scale(demand = self$unresolved(), throttle = throttle)
    self$tally()
    daemons %|||% daemons_info(name = private$.name, seconds_interval = priv...
   daemons_info(name = private$.name, seconds_interval = private$.seconds_i...
    crew_retry(fun = ~{ daemons <- mirai::status(.compute = name)$daemons va...
    isTRUE(do.call(what = fun, args = args, envir = envir))
    do.call(what = fun, args = args, envir = envir)
    (function (..., .x = ..1, .y = ..2, . = ..1) { daemons <- mirai::status(...
    mirai::status(.compute = name)
    query_status(envir)
# ...
```

We're just stumped and can't tell if it's a bug in our pipeline or it's some sort of parallelization issue that we're having. We were also started receiving this error, but strangely were able to fix it by setting `launch_max = 100`.

```r
    {crew} worker 15 launched 5 times in a row without completing any tasks. Either troubleshoot or raise launch_max asLast error traceback:
```

**Comment by MilesMcBain** [comment]
It looks like this *could* be coming from somewhere in targets, it's hard to tell.

The way to find out for sure is use the `callr_function = NULL` method outlined in [the manual](https://books.ropensci.org/targets/debugging.html) such that the entire pipeline runs within your interactive session and you'll be able to use all R's debugging tools.

In this context I am a fan of using `options(error = recover)` to pause evaluation when the error is hit, letting me poke around at various bits of data to see what the problem is. If you've never used `recover` before, I can say it is WELL WORTH the investment to figure it out, possibly only second to `{targets}` itself in this regard.

**Reply by dsweber2** [reply]
That is eventually on the list of things for us to try, but narrowing down which of ~64 jobs that take ~an hour each makes it difficult to actually run `tar_make(this_job_is_definitely_the_problem, callr_function = NULL)` to be able to debug it. So far the serial run has worked fine, and any of our versions with abridged targets also were working fine, so we're stuck with more obscure methods.

We were hoping that `received data could not be converted to integer` and some of the traceback were more clear to someone with more knowledge about the targets/crew internals, but hopefully we'll be able to get a clearer picture of the error in the coming weeks.

**Reply by MilesMcBain** [reply]
Do any of the dispatched targets complete before you see the error?

If not I am thinking, and I could be way off, it has the feel of some kind of serialisation issue, potentially when trying to ship the data from the worker back to the main thread to be saved.

If it were my pipeline I'd try the `tar_option_set(storage = "worker")` as a band-aid. But I'd still want to understand what was failing to serialise and why.

**Reply by shikokuchuo** [reply]
> received data could not be converted to integer

This is a low level error from {nanonext}, a dependency of {mirai} and {crew}.

This points to a synchronization issue with the parallel workers. As this is the first time it has ever appeared for us (the developers), this leads me to suspect that this is a recent regression. For {mirai} 0.12.1 we did make some code simplifications in an area that could possibly trigger this. This is reverted in the current development build mirai 0.13.0.9001.

If you are able to try the current development builds using the R-universe repo below – if that solves the problem for you then we will make sure this makes it in to the next release. Thanks.

```r
install.packages(c('nanonext', 'mirai'), repos = 'shikokuchuo.r-universe.dev')
```

@milesmcbain thanks for trying to help – you were very close to the truth, and your suggestion makes good sense in the context.

**Reply by dsweber2** [reply]
thank you both for the suggestions! 
So with the following options (which we used and got the above error):
```
main_controller <- crew_controller_local(
  name = "main_controller",
  workers = parallel::detectCores() - 1L,
  seconds_idle = 60L,
  tasks_max = 1L,
  launch_max = 100L
)
```
After installing the dev versions of nanonext and mirai, we're back to getting 
```
    {crew} worker 42 launched 100 times in a row without completing any tasks. Either troubleshoot or raise launch_max above 100. Details: https://wlandau.github.io/crew/articles/risks.html#crashes
```
which is at least more interpretable.

**Reply by dsweber2** [reply]
I think we're also getting way more R processes spawning than we have workers; `parallel::detectCores() - 1L=63`, but when I look at `ps -e | grep R | wc -l` we have 135 lines (1-4 of these may be unrelated, but it's over double the count!). I was expecting to see ~ the same number of R processes as the intended worker count, though this may just be a lack of familiarity with the multi-processing used here.

**Reply by shikokuchuo** [reply]
Thanks for reporting back @dsweber2. I am glad you are no longer seeing the `received data could not be converted to integer` error. We will maintain this fix in the main branch going forward. I believe @wlandau has a set of guidance for errors of the type you are getting.

@wlandau FYI the fix in {mirai} reverts the dispatcher control sockets to req/rep - although we were not able to break it in testing, it evidently has in the wild. Although it may only be a symptom of a connection error, the low level error is not easy for you to handle, so this is the best course of action to ensure robustness going forward.

**Reply by wlandau** [reply]
Yes, glad the error is more interpretable. A more complete reprex would help diagnose the worker crash problem: https://books.ropensci.org/targets/help.html. You could also turn on local logs, c.f. `local_log_directory`: https://wlandau.github.io/crew/reference/crew_controller_local.html

**Reply by wlandau** [reply]
If there are still issues with `crew` workers, please open a new thread and include a small reprex.

**Reply by dshemetov** [reply]
It's been tough to isolate the issue on our end, but we'll keep you posted.
